server/insertlist_test.go

```diff
@@ -0,0 +1,181 @@
+package server
+
+import (
+	"bytes"
+	"errors"
+	"fmt"
+	"net/http"
+	"net/http/httptest"
+	"testing"
+
+	"github.com/TheYeung1/yata-server/database"
+	"github.com/TheYeung1/yata-server/model"
+	"github.com/TheYeung1/yata-server/server/request"
+	"github.com/stretchr/testify/assert"
+)
+
+func TestInsertListInput_Validate(t *testing.T) {
+	tests := []struct {
+		input InsertListInput
+		err   error
+	}{
+		{
+			input: InsertListInput{ListID: "ID", Title: "Title"},
+		},
+		{
+			input: InsertListInput{ListID: ""},
+			err:   errors.New("ListID cannot be empty"),
+		},
+		{
+			input: InsertListInput{ListID: "Lorem ipsum dolor sit amet, consectetur adipiscing elit. Pellentesque porta eros erat. Curabitur nam."},
+			err:   errors.New("ListID length cannot exceed 100 characters"),
+		},
+		{
+			input: InsertListInput{ListID: "ID "},
+			err:   errors.New("ListID cannot be prefixed or suffixed with spaces"),
+		},
+		{
+			input: InsertListInput{ListID: " ID"},
+			err:   errors.New("ListID cannot be prefixed or suffixed with spaces"),
+		},
+		{
+			input: InsertListInput{ListID: "ID"},
+			err:   errors.New("Title cannot be empty"),
+		},
+		{
+			input: InsertListInput{ListID: "ID", Title: "Lorem ipsum dolor sit amet, consectetur adipiscing elit. Pellentesque porta eros erat. Curabitur nam."},
+			err:   errors.New("Title length cannot exceed 100 characters"),
+		},
+		{
+			input: InsertListInput{ListID: "ID", Title: "Title "},
+			err:   errors.New("Title cannot be prefixed or suffixed with spaces"),
+		},
+		{
+			input: InsertListInput{ListID: "ID", Title: " Title"},
+			err:   errors.New("Title cannot be prefixed or suffixed with spaces"),
+		},
+	}
+
+	for _, test := range tests {
+		test := test
+		t.Run(fmt.Sprintf("%+v", test.input), func(t *testing.T) {
+			t.Parallel()
+			assert.Equal(t, test.err, test.input.Validate())
+		})
+	}
+}
+
+func TestServer_InsertList(t *testing.T) {
+	tests := []struct {
+		input      string
+		insertList func(*testing.T) func(id model.UserID, list model.YataList) error
+		outCode    int
+		outBody    string
+	}{
+		{
+			input: "{\"ListID\":\"ID\",\"Title\":\"Title\"}",
+			insertList: func(t *testing.T) func(id model.UserID, list model.YataList) error {
+				return func(id model.UserID, list model.YataList) error {
+					assert.Equal(t, model.UserID("u"), id)
+					assert.Equal(t, model.YataList{
+						UserID: "u",
+						ListID: "ID",
+						Title:  "Title",
+					}, list)
+					return nil
+				}
+			},
+			outCode: http.StatusCreated,
+			outBody: "{\"ListID\":\"ID\"}\n",
+		},
+	}
+
+	for _, test := range tests {
+		test := test
+		t.Run(fmt.Sprintf("%+v", test.input), func(t *testing.T) {
+			t.Parallel()
+
+			rec := httptest.NewRecorder()
+			req := httptest.NewRequest( /* Method */ "", "https://does.not/matter", bytes.NewBufferString(test.input))
+
+			srvr := Server{Ydb: MockTdb{MockInsertList: test.insertList(t)}}
```

Comment:
With the function receiver approach you have to create an entire `Server` just to test a single handler. You also need to create the mock database (tho you only need to define that once for all tests so not a big deal)

Reply:
I'm not entirely sure one is more idiomatic than the other. One of the things I really like about the second method is it makes it *very* clear what the dependencies of a particular handler are. With the server-based approach you don't know, without clicking into the code, what parts, if any, of the server the hander is using.